Version 2.0.0

markbao opened the pull request:
## 2.0 Checklist

- [x] Merge in @mikepb's passcode fork
- [x] Identify areas of API breakage
- [x] Think about whether we want to make breaking changes, return to original API, or display a deprecation notice for 2.0 and remove later
- [x] Consider adding a deprecation notice
- [x] Write out full README with guide
- [x] Think about how to do a demo ([Tonic](https://tonicdev.uservoice.com/knowledgebase/articles/765846-how-do-i-customize-the-example-for-my-npm-package)?)
- [x] Remove console.log debug artifacts
- [x] Write tests for `generate_key` and `generate_key_ascii`
- [x] Run JSHint (or consider using Semistandard and use the auto-formatter)
- [x] Edit and move over documentation
- [x] Check tests
- [x] Check issues
- [x] Add `.counter` and `.time` back in for API compatibility
- [x] Last check for API breakage
- [x] Update HISTORY.md

## API breakage

- [x] Options for hotp and totp: `key` is now `secret`
- [x] Options for totp: `initial_time` is now `epoch`
- [x] `generate_key` method missing

## Open questions

- [x] Move `verify()` to `verifyDelta()` and create a new `verify()` that returns a boolean for a strict comparison (delta = 0)
- [x] `generate_key` -> `generateSecret()` for consistency?
- [x] Do we still want to do `windowAhead` and `windowBehind`?

mikepb commented on README.md:
```diff
@@ -1,160 +1,53 @@
+## Documentation
 
-You'll need to have `vows` installed. If you don't have vows, `npm install -g vows`
+Full documentation at http://mikepb.github.io/passcode/
```
Need new URL?

mikepb replied:
Just saw the in-progress in the title. I used `npm run doc` to generate the HTML docs, but I'm not sure if that still works.

markbao replied:
Will give that a shot. Join me here? https://gitter.im/speakeasyjs/speakeasy

markbao commented:
@mikepb It would be interesting to have a verifyTOTP function that returns a bool. What do you think? Any idea what to name it?

mikepb commented:
I think that pattern was from one of the other OTP modules I "borrowed" from. How about `verifyDelta()` that returns the object and `verify()` that wraps `verifyDelta()` to return a boolean?

markbao commented:
Great idea. Added it to the list.

mikepb commented:
I like the new logo btw!

coolaj86 commented:
I'm all for writing code for the fun of it if that's what the goal is here, but there's already the [`notp`](https://github.com/guyht/notp) module that is very simple and very node-esque. Why not just use that?

I did just that with [`authenticator`](https://github.com/Daplie/node-authenticator) and you could probably copy that code and create a backwards compatibility layer on top `notp` very easily.

mikepb commented:
I wrote the `passcode` library because I was not satisfied with the (then) available One-Time Password libraries. In fact, passcode (on which V2 of speakeasy is based) incorporates the design, code, and tests of `notp` and ideas from other OTP implementations. It's beside the point that `notp` exists as a sufficient module. Open source software benefits from every contribution, regardless of purpose.

markbao commented:
@coolaj86 Hey AJ, definitely hear where you're coming from here. notp is very good, and it is plausible to use notp as the base for Speakeasy, but this package does have a somewhat larger API than notp and thus would require a lot of updates to notp.

When considering what the 2.0 update would look like, I strongly considered your suggestion in another issue to base it on notp, but it would have required extending notp quite a bit – like allowing for custom key lengths, adding the key generator, adding epochs (initial time) for TOTP, and stuff like that – that it made the most sense (to me) that this package should be standalone.

Of course, this means more work, but I think it's the best way to go and hopefully offers more choice to the Node community. Thanks for offering to provide `authenticator` as a base, by the way!

markbao commented on index.js:
```diff
@@ -1,2 +1,544 @@
+  if (options.counter == null) options.counter = exports._counter(options);
+
+  // adjust for two-sided window
+  options.counter -= window;
```
@mikepb Do we just need to add this one adjustment for both sides of the window?

mikepb replied:
Yep, this adjustment is to allow reuse of the `hotp.verifyDelta()` function.

Before range: [counter, counter+window]
After adjustment: [counter-window, counter+window]

markbao commented:
Alright, I think this is pretty much there.

- I've done automatic and manual checks to make sure that the Speakeasy 1.0 API is compatible with the 2.0 API (albeit with some deprecations).
- From running the 1.0 tests, I found that `time` needs to be set in seconds (as in 1.0), not milliseconds (as in previous commits in 2.0).
- Renamed `googleAuthURL()` to `otpauth_url()` to be more general. 
- Ran Semistandard (Standard with semicolons) to lint the codebase and it's now passing.
- Added documentation for how to use Speakeasy for two-factor authentication.
- Lots of other stuff.

All in all, I think this is good to go. @mikepb, mind giving this a review?

mikepb commented on History.md:
```diff
@@ -1,30 +0,0 @@
-Speakeasy
```
Copy this history over to 2.0?

markbao replied:
Added back in in cd14b0f0b648643f97ea9b9a543fbc4ed44da35c.

mikepb commented on README.md:
```diff
@@ -1,160 +1,529 @@
+
+// Display this data URL to the user in an <img> tag
+// Example:
+write('<img src="' + data_url + '">');
```
This line should go inside the function closure above. For async, `data_url` isn't guaranteed to have a value here (actually, it's not even defined in this context).

markbao replied:
Yeah, I was trying to illustrate a multi-step process as if these pieces of code would be in different places, but obviously it doesn't make sense as a single block of code. Updated it to put this line in the callback in 37d4937122cfcd4cb96f64db3e371dfe424a4691

mikepb commented on asdf.md:
```diff
@@ -0,0 +1,336 @@
+## Functions
```
asdf?

markbao replied:
Whoops, that's the interim file that jsdoc2md generates that I forgot to remove. Deleted.

mikepb commented on index.js:
```diff
@@ -1,2 +1,601 @@
+  var hmac = crypto.createHmac(algorithm, key);
+
+  // update hmac with the counter
+  hmac.update(buf);
```
Regarding issue #35, while implementing passcode, one of the HOTP libraries repeated the secret to the length of the algorithm. I don't think I found any information on this behavior in the RFC's. I experimented with both and ultimately decided to not repeat the secret. We should check on this behavior before publishing version 2.0.

markbao replied:
Got it. I wonder if there's anyone we can ask that would know more about this.

mikepb replied:
Last time I just dug through the google authenticator code to find out. I don't remember seeing code to repeat the secret up to the hash size. 

> On Jan 24, 2016, at 10:00 AM, Mark Bao <notifications@github.com> wrote:
> 
> In index.js:
> 
> > +  // create an buffer from the counter
> > +  var buf = new Buffer(8);
> > +  var tmp = counter;
> > +  for (i = 0; i < 8; i++) {
> > +    // mask 0xff over number to get last 8
> > +    buf[7 - i] = tmp & 0xff;
> > +
> > +    // shift 8 and get ready to loop over the next batch of 8
> > +    tmp = tmp >> 8;
> > +  }
> > +
> > +  // init hmac with the key
> > +  var hmac = crypto.createHmac(algorithm, key);
> > +
> > +  // update hmac with the counter
> > +  hmac.update(buf);
> Got it. I wonder if there's anyone we can ask that would know more about this.
> 
> —
> Reply to this email directly or view it on GitHub.
>

markbao replied:
Weird. I guess one way to figure it out is to test with a short secret on 256 and 512 with GA.

mikepb replied:
None of these implementations repeat the key:
- https://github.com/google/google-authenticator/blob/master/mobile/blackberry/src/com/google/authenticator/blackberry/PasscodeGenerator.java#L81
- https://github.com/gokyle/hotp/blob/master/hotp.go#L238
- https://github.com/google/google-authenticator/blob/master/mobile/ios/Classes/OTPGenerator.m#L102
- https://github.com/Jakobo/hotp-php/blob/master/hotp.php#L18
- https://github.com/adulau/hotp-js/blob/master/hotp/hotp.js#L89

Though I found a reference to padding for IPSec:
- http://www.rfc-archive.org/getrfc.php?rfc=4868

mikepb commented:
I did a quick review this morning, but will have more time this evening for a fuller review.

mikepb commented on index.js:
```diff
@@ -1,2 +1,601 @@
+  }
+
+  // generate an ascii key
+  var key = this.generate_key_ascii(length, symbols);
```
This is deprecated :wink:

markbao replied:
Ha, good catch! Fixed.

mikepb commented on README.md:
````diff
@@ -1,160 +1,529 @@
-* `encoding` (default `ascii`): the encoding of the `key`. Can be `'ascii'`, `'hex'`, or `'base32'`. The key will automatically be converted to ASCII.
+```js
+// Let's say the user says that the token they have is 132890
+userToken = 132890;
````
This should be a string (from a HTML form input)

markbao replied:
Ah good catch, fixed in f782374.

mikepb commented on README.md:
````diff
@@ -1,160 +1,529 @@
-## Issues and patches
+```js
+var token = speakeasy.hotp({
+  secret: base32secret,
````
`secret.base32`

I would like to make it clear that this value is to be persisted to a database, e.g. `db.fetchSecretForUser()`. This was not completely clear to me when I first looked at TOTP.

markbao replied:
Ha, that's actually exactly what I did but I removed it, but forgot to remove base32secret. In step 1, I had:

```js
current_user.two_factor_temp_secret = secret.base32
```

And step 3 had:

```js
base32secret = current_user.two_factor_temp_secret
```

But I removed all of that because I was worried it would be confusing, and replaced it with the text that it should be stored somewhere temporarily. Not sure what the right way to do it is.

mikepb commented on index.js:
```diff
@@ -1,2 +1,601 @@
+  // unpack digits
+  // backward compatibility: `length` is also accepted here, but deprecated
+  var digits = (options.digits != null ? options.digits : options.length) || 6;
+  if (options.length) console.log('Speakeasy - Deprecation Notice - Specifying token digits using `length` is no longer supported. Use `digits` instead.');
```
`options.length != null` to check that the option is `undefined` or `null`. Here, `length=0` matches, though that's a different error.

markbao replied:
I did `if (options.length)` as a shortcut since 0 would also be an error, but that's true – 0 doesn't really fall under this. I'll have to do something like `if (options.length !== null || typeof options.length !== 'undefined')` since Semistandard enforces `===` and `!==`.

mikepb replied:
`== null` actually matches both `null` and `undefined` but not integers or other values

```js
$ node
> var x = 0;
undefined
> x == null
false
> var y;
undefined
> y
undefined
> y == null
true
```

markbao replied:
Issue here though is that Semistandard requires `===` and will error if given a `==`:

```js
> var y
undefined
> y == null
true
> y === null
false
```

mikepb replied:
I see. I guess I'm not as pedantic about equality as Semistandard...

markbao replied:
My bad, fixed in a7230cb.

mikepb commented on README.md:
````diff
@@ -1,160 +1,529 @@
-* `encoding` (default `ascii`): the encoding of the `key`. Can be `'ascii'`, `'hex'`, or `'base32'`. The key will automatically be converted to ASCII.
+```js
+// Let's say the user says that the token they have is 132890
+userToken = '132890';
````
Global leak! `var userToken = '132890';`

markbao replied:
/headdesk 87da5f8

mikepb commented on index.js:
```diff
@@ -1,2 +1,601 @@
+ * @global
+ */
+exports.hotp.verify = function hotpVerify (options) {
+  return exports.hotp.verifyDelta(options) != null;
```
Semistandard complained here?

markbao replied:
Actually, I'm wrong. I remember it complained because of an equality with `null` (see below) but it seems to only complain with `x == null`, not `x != null`. Sorry about that.

https://github.com/speakeasyjs/speakeasy/commit/a8c306ff14018cfd39ac59779ef5eadcd57c5213#diff-168726dbe96b3ce427e7fedce31bb0bcL551

mikepb replied:
Seems inconsistent to do that

markbao replied:
Yeah, agreed. Not sure why it does that.

mikepb commented on index.js:
```diff
@@ -1,2 +1,601 @@
+ * @param {Object} options
+ * @param {String} options.secret Shared secret key
+ * @param {String} options.token Passcode to validate
+ * @param {Integer} [options.time] Time with which to calculate counter value.
```
Time in seconds... (and all other occurrences)

markbao replied:
Thanks, missed that one and a few ones for `epoch` for some reason. f12bd0a

mikepb commented on index.js:
```diff
@@ -1,2 +1,601 @@
+  // loop from C to C + W
+  for (i = counter; i <= counter + window; ++i) {
+    options.counter = i;
+    if (exports.hotp(options) === token) {
```
Should we do constant-time comparison here, for security purposes?

markbao replied:
How so?

mikepb replied:
A variable time string comparison could leak information about common prefixes because these comparisons take less time. Though the randomization used in HOTP would mitigate against this and it would take a large number of comparisons to get any information out of the system (unlikely if repeated failed attempts are blocked). In short, I would guess this attack would be impractical.

https://wiki.php.net/rfc/timing_attack

Regardless, we could still implement constant-time token comparison efficiently based on the fact that the strings are just digits.

```js
var token = String(token);
var digits = parseInt(options.digits || 0, 10);

// fail if token is not of correct length
if (token.length !== digits) {
  return;
}

// parse token to integer
token = parseInt(token, 10);

// loop from C to C + W inclusive
for (i = counter; i <= counter + window; ++i) {
  options.counter = i;
  // domain-specific constant-time comparison for integer codes
  if (parseInt(exports.hotp(options), 10) === token) {
    // found a matching code, return delta
    return {delta: i - counter};
  }
}
```

markbao replied:
Got it. Are there any issues you would think would arise from comparing converted ints instead of strings? I can't think of any right now, considering that front zero-padding shouldn't make a difference.

mikepb replied:
Not off the top of my head, assuming that `parseInt(•, 10)` doesn’t have any major flaws.

> On Jan 24, 2016, at 11:58 AM, Mark Bao <notifications@github.com> wrote:
> 
> In index.js <https://github.com/speakeasyjs/speakeasy/pull/44#discussion_r50641088>:
> 
> > +
> > +exports.hotp.verifyDelta = function hotpVerifyDelta (options) {
> > +  var i;
> > +
> > +  // shadow options
> > +  options = Object.create(options);
> > +
> > +  // unpack options
> > +  var token = options.token;
> > +  var window = parseInt(options.window || 0, 10);
> > +  var counter = parseInt(options.counter || 0, 10);
> > +
> > +  // loop from C to C + W
> > +  for (i = counter; i <= counter + window; ++i) {
> > +    options.counter = i;
> > +    if (exports.hotp(options) === token) {
> Got it. Are there any issues you would think would arise from comparing converted ints instead of strings? I can't think of any right now, considering that front zero-padding shouldn't make a difference.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/speakeasyjs/speakeasy/pull/44/files#r50641088>.
>

markbao replied:
Got it. Mind writing this up and making a PR for it?

mikepb replied:
#49

mikepb commented on index.js:
```diff
@@ -1,2 +1,601 @@
+  // unpack options
+  var token = options.token;
+  var window = parseInt(options.window || 0, 10);
+  var counter = parseInt(options.counter || 0, 10);
```
Just occurred to me that the default values should be provided after parsing, otherwise, we get a `NAN`.

```js
parseInt(options.counter, 10) || 0;
```

```
> parseInt("d", 10)
NaN
> parseInt("d", 10) || 6
6
```

markbao replied:
Fixed in f653975.

markbao commented on index.js:
```diff
@@ -1,2 +1,601 @@
+ *   difference between the client and the server as the `delta` property (i.e.
+ *   `{ delta: 0 }`).
+ * @method hotp․verifyDelta
+ * @global
```
@mikepb, any reason these are defined as global?

mikepb replied:
JSDoc wouldn’t play nice if I didn’t...

> On Jan 25, 2016, at 10:11 AM, Mark Bao <notifications@github.com> wrote:
> 
> In index.js <https://github.com/speakeasyjs/speakeasy/pull/44#discussion_r50731507>:
> 
> > + *   the application and must be incremented for each request.
> > + * @param {Integer} [options.digits=6] The number of digits for the one-time
> > + *   passcode.
> > + * @param {Integer} [options.window=0] The allowable margin for the counter.
> > + *   The function will check "W" codes in the future against the provided
> > + *   passcode, e.g. if W = 10, and C = 5, this function will check the
> > + *   passcode against all One Time Passcodes between 5 and 15, inclusive.
> > + * @param {String} [options.encoding="ascii"] Key encoding (ascii, hex,
> > + *   base32, base64).
> > + * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
> > + *   sha512).
> > + * @return {Object} On success, returns an object with the counter
> > + *   difference between the client and the server as the `delta` property (i.e.
> > + *   `{ delta: 0 }`).
> > + * @method hotp․verifyDelta
> > + * @global
> @mikepb <https://github.com/mikepb>, any reason these are defined as global?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/speakeasyjs/speakeasy/pull/44/files#r50731507>.
>

markbao replied:
Excellent reason.

mikepb commented:
:+1: :rocket: hopefully no :bug:

markbao commented:
@mikepb I'm having a bit of an issue testing a TOTP window.

```js
var secret = 'rNONHRni6BAk7y2TiKrv'

// Get current TOTP value
> speakeasy.totp({secret: secret});
'867785'

// Get delta with varying window lengths
> speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:0});
{ delta: 0 }
> speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:1});
{ delta: 1 }
> speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:2});
{ delta: 2 }
> speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:3});
{ delta: 3 }
```

This might be an implementation detail, but it seems like we are calculating the TOTP window from the beginning of the window even though it's a two-sided window. I suppose I expected the deltas for these to all be `0`, since they are `0` away from the given token. But it seems like they are `window` away from the start of the window.

mikepb commented:
Looks like that's a side effect from the window adjustment to allow for reuse of hotp.verify. I never used the delta feature, so never occurred to me to write tests for it. 

Good catch!

> On Jan 26, 2016, at 10:10 AM, Mark Bao <notifications@github.com> wrote:
> 
> @mikepb I'm having a bit of an issue testing a TOTP window.
> 
> var secret = 'rNONHRni6BAk7y2TiKrv'
> 
> // Get current TOTP value
> > speakeasy.totp({secret: secret});
> '867785'
> 
> // Get delta with varying window lengths
> > speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:0});
> { delta: 0 }
> > speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:1});
> { delta: 1 }
> > speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:2});
> { delta: 2 }
> > speakeasy.totp.verifyDelta({secret: secret, token:'867785', window:3});
> { delta: 3 }
> This might be an implementation detail, but it seems like we are calculating the TOTP window from the beginning of the window even though it's a two-sided window. I suppose I expected the deltas for these to all be 0, since they are 0 away from the given token. But it seems like they are window away from the start of the window.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mikepb commented:
Pull #50 fixes the TOTP window bug.

markbao commented on index.js:
```diff
@@ -1,2 +1,624 @@
+        throw new Error('Speakeasy - otpauthURL - Invalid digits `' + digits + '`');
+    }
+    query.digits = digits;
+  }
```
@mikepb, do you think we need to be restricting the `digits` here if it's a generic otpauth:// generator? I haven't been able to find any sort of spec for otpauth:// other than the one in the Google Authenticator wiki.

mikepb replied:
No reason other than Google, but we could just make a warning instead of full-blown error.

> On Jan 26, 2016, at 8:51 PM, Mark Bao <notifications@github.com> wrote:
> 
> In index.js <https://github.com/speakeasyjs/speakeasy/pull/44#discussion_r50941080>:
> 
> > +        throw new Error('Speakeasy - otpauthURL - Invalid algorithm `' + algorithm + '`');
> > +    }
> > +    query.algorithm = algorithm.toUpperCase();
> > +  }
> > +
> > +  // validate digits
> > +  if (digits != null) {
> > +    switch (parseInt(digits, 10)) {
> > +      case 6:
> > +      case 8:
> > +        break;
> > +      default:
> > +        throw new Error('Speakeasy - otpauthURL - Invalid digits `' + digits + '`');
> > +    }
> > +    query.digits = digits;
> > +  }
> @mikepb <https://github.com/mikepb>, do you think we need to be restricting the digits here if it's a generic otpauth:// generator? I haven't been able to find any sort of spec for otpauth:// other than the one in the Google Authenticator wiki.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/speakeasyjs/speakeasy/pull/44/files#r50941080>.
>

markbao replied:
Good call. Should we do the same for algorithm as well?

mikepb replied:
Yep.

> On Jan 26, 2016, at 9:05 PM, Mark Bao <notifications@github.com> wrote:
> 
> In index.js <https://github.com/speakeasyjs/speakeasy/pull/44#discussion_r50941798>:
> 
> > +        throw new Error('Speakeasy - otpauthURL - Invalid algorithm `' + algorithm + '`');
> > +    }
> > +    query.algorithm = algorithm.toUpperCase();
> > +  }
> > +
> > +  // validate digits
> > +  if (digits != null) {
> > +    switch (parseInt(digits, 10)) {
> > +      case 6:
> > +      case 8:
> > +        break;
> > +      default:
> > +        throw new Error('Speakeasy - otpauthURL - Invalid digits `' + digits + '`');
> > +    }
> > +    query.digits = digits;
> > +  }
> Good call. Should we do the same for algorithm as well?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/speakeasyjs/speakeasy/pull/44/files#r50941798>.
>

markbao commented:
Alright, here we go!

markbao commented:
And we're live! https://www.npmjs.com/speakeasy

Thank you to all of the contributors to version 2.0.0.

Super thanks to @mikepb – I really enjoyed working with you on this, from having Passcode as a solid foundation to build on, to your input into design decisions and making pull requests. Thank you for your invaluable help on this project.

Really proud to have this out!

![](http://i.giphy.com/102VfCWF40oAuI.gif)

mikepb commented:
:clap:
